cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java

```diff
@@ -403,6 +467,70 @@ public void run() {
     }
   }
 
+  /**
+   * Serializes a {@link CredentialAccessBoundary} object into Protobuf wire format.
+   */
+  private byte[] serializeCredentialAccessBoundary(
+      CredentialAccessBoundary credentialAccessBoundary) throws IOException {
+    List<AccessBoundaryRule> rules =
+        credentialAccessBoundary.getAccessBoundaryRules();
+    ClientSideAccessBoundary.Builder accessBoundaryBuilder =
+        ClientSideAccessBoundary.newBuilder();
+
+    for (AccessBoundaryRule rule : rules) {
+      ClientSideAccessBoundaryRule.Builder ruleBuilder =
+          accessBoundaryBuilder.addAccessBoundaryRulesBuilder()
+              .addAllAvailablePermissions(rule.getAvailablePermissions())
+              .setAvailableResource(rule.getAvailableResource());
+
+      if (rule.getAvailabilityCondition() != null) {
+        String availabilityCondition =
+            rule.getAvailabilityCondition().getExpression();
+
+        Expr availabilityConditionExpr = this.compileCel(availabilityCondition);
+        ruleBuilder.setCompiledAvailabilityCondition(availabilityConditionExpr);
+      }
+    }
+
+    return accessBoundaryBuilder.build().toByteArray();
+  }
+
+  /**
+   * Compiles CEL expression from String to an {@link Expr} proto object. 
+   */
+  private Expr compileCel(String expr) throws IOException {
+    try {
+      CelAbstractSyntaxTree ast = celCompiler.parse(expr).getAst();
+
+      CelProtoAbstractSyntaxTree astProto =
+          CelProtoAbstractSyntaxTree.fromCelAst(ast);
+
+      return astProto.getExpr();
+    } catch (CelValidationException exception) {
+      throw new IOException("Failed to parse CEL expression: " +
+                            exception.getMessage());
```

Comment:
Why re-throw as IOException since it doesn't seem like an I/O occurs here? Are we trying to hide the underlying CEL logic from the users?

Reply:
Not really. Changed it to throw CelValidationException directly